### PR TITLE
feat: /dockerhub/tag/:owner/:image

### LINKED ISSUE #6916: Get feed about latest tags in docker hub
### What feature is it?
Get info about new tags

### What problem does this feature solve?
Last tag info (11 feb 2021)
image

- What is actually happening
There info about old tag (12 may 2017), not about new 
image
image

### PR DESCRIPTION
**Pull request by outloudvi**
## 该 PR 相关 Issue / Involved issue

Close #6916

## 完整路由地址 / Example for the proposed route(s)

```routes
/dockerhub/tag/library/mariadb
/dockerhub/tag/rustembedded/cross
```

## 新RSS检查列表 / New RSS Script Checklist
  
- [x] New Route
- [x] Documentation
  - [x] CN
  - [x] EN
- [ ] 全文获取 fulltext
  - [ ] Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [x] 日期和时间 date and time
  - [x] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added 
- [ ] `Puppeteer`

## 说明 / Note

将来也许可以将 architecture 作为过滤参数以只选择支持特定架构的 tag。

### CODE REVIEW

File: lib/v2/dockerhub/tag.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        title: `${namespace} tags`,
+        description: metadata.data.description,
+        link,
+        language: 'en',
+        item: tags.map((item) => ({
+            title: `${namespace}:${item.name} was updated`,
+            description: `${namespace}:${item.name} was updated, supporting the architectures of ${item.images.map((img) => `${img.os}/${img.architecture}`).join(', ')}`,
+            link: `https://hub.docker.com/layers/${owner === 'library' ? `${image}/` : ''}${namespace}/${item.name}/images/${item.images[0].digest.replace(':', '-')}`,
+            author: owner,
+            pubDate: new Date(item.tag_last_pushed).toUTCString(),
```
Comment (TonyRL) [inline lib/v2/dockerhub/tag.js:25]:
```suggestion
            pubDate: parseDate(item.tag_last_pushed),
```
Please use parseDate() https://github.com/DIYgod/RSSHub/issues/6795

File: lib/v2/dockerhub/radar.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+    'hub.docker.com': {
+        _name: 'Docker Hub',
+        '.': [
+            {
+                title: '镜像有新 Tag',
+                docs: 'https://docs.rsshub.app/program-update.html#docker-hub-jing-xiang-you-xin-tag',
```
Comment (TonyRL) [inline lib/v2/dockerhub/radar.js:7]:
```suggestion
                docs: 'https://docs.rsshub.app/program-update.html#docker-hub',
```
This should point to H2 of the docs 
See https://docs.rsshub.app/joinus/quick-start.html#ti-jiao-xin-de-rsshub-radar-gui-ze-bian-xie-gui-ze-docs

File: lib/radar-rules.js [hunk 3]
```diff
@@ -791,7 +791,7 @@ module.exports = {
         hub: [
             {
                 title: '镜像有新 Build',
-                docs: 'https://docs.rsshub.app/program-update.html#docker-hub',
+                docs: 'https://docs.rsshub.app/program-update.html#docker-hub-jing-xiang-you-xin-build',
```
Comment (TonyRL) [inline lib/radar-rules.js:794]: Please remove old entry in radar-rules.js

**Review by TonyRL — CHANGES_REQUESTED** [review]

File: lib/radar-rules.js (hunk 3 above)
Comment (outloudvi) [reply lib/radar-rules.js:794]: The old entry is not in `v2`. Shall I migrate it there?

Comment (TonyRL) [reply lib/radar-rules.js]: Yes.
See https://github.com/DIYgod/RSSHub/issues/8269

File: lib/v2/dockerhub/radar.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module.exports = {
+    'hub.docker.com': {
+        _name: 'Docker Hub',
+        '.': [
```
Comment (TonyRL) [inline lib/v2/dockerhub/radar.js:4]:
```suggestion
    'docker.com': {
        _name: 'Docker Hub',
        'hub': [
```

File: docs/en/program-update.md [hunk 5]
```diff
@@ -119,6 +119,16 @@ For example: `https://www.curseforge.com/sc2/assets/taylor-mouses-stuff/files` t
...
+### Image New Tag
+
+<RouteEn author="outloudvi" example="/dockerhub/tag/library/mariadb" path="/dockerhub/tag/:owner/:image" :paramsDesc="['Image owner', 'Image name']" />
+
+::: warning
+
+Use `library` as the `owner` for official images, such as <https://rsshub.app/dockerhub/tag/library/mysql>
+
+:::
+
```
Comment (TonyRL) [inline docs/en/program-update.md:131]:
```suggestion
<RouteEn author="outloudvi" example="/dockerhub/tag/library/mariadb" path="/dockerhub/tag/:owner/:image" :paramsDesc="['Image owner', 'Image name']">

::: warning

Use `library` as the `owner` for official images, such as <https://rsshub.app/dockerhub/tag/library/mysql>

:::

</RouteEn>
```
Please put the warning inside <RouteEn> tag

File: docs/program-update.md [hunk 6]
```diff
@@ -158,6 +158,16 @@ pageClass: routes
...
+### 镜像有新 Tag
+
+<Route author="outloudvi" example="/dockerhub/tag/library/mariadb" path="/dockerhub/tag/:owner/:image" :paramsDesc="['镜像作者', '镜像名称']" radar="1" rssbud="1"/>
+
+::: warning 注意
+
+官方镜像的 owner 填写 library, 如: <https://rsshub.app/dockerhub/tag/library/mysql>
+
+:::
+
```
Comment (TonyRL) [inline docs/program-update.md:170]:
```suggestion
<Route author="outloudvi" example="/dockerhub/tag/library/mariadb" path="/dockerhub/tag/:owner/:image" :paramsDesc="['镜像作者', '镜像名称']" radar="1" rssbud="1">

::: warning 注意

官方镜像的 owner 填写 library, 如: <https://rsshub.app/dockerhub/tag/library/mysql>

:::

</Route>

```
Please put the warning inside <Route> tag

**Review by TonyRL — CHANGES_REQUESTED** [review]

File: lib/v2/dockerhub/tag.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        description: metadata.data.description,
+        link,
+        language: 'en',
+        item: tags.map((item) => ({
+            title: `${namespace}:${item.name} was updated`,
+            description: `${namespace}:${item.name} was updated, supporting the architectures of ${item.images.map((img) => `${img.os}/${img.architecture}`).join(', ')}`,
+            link: `https://hub.docker.com/layers/${owner === 'library' ? `${image}/` : ''}${namespace}/${item.name}/images/${item.images[0].digest.replace(':', '-')}`,
+            author: owner,
+            pubDate: parseDate(item.tag_last_pushed),
+            guid: item.tag_last_pushed,
```
Comment (TonyRL) [inline lib/v2/dockerhub/tag.js:27]: Consider not setting the guid.
If guid is not set, it will use `link` instead, which has a SHA256 that is more collision resistance than a timestamp imo.

**Comment by TonyRL** [comment]
Consider adding an optional `page_size` parameter to the API endpoint as some image(OpenJDK) has a lot of tags.
New tags may not appear within the default value of 10.

**Comment by outloudvi** [comment]
> Consider adding an optional page_size parameter to the API endpoint as some image(OpenJDK) has a lot of tags.

Sounds great. By the way, is there any policy here on the paginator?

File: lib/v2/dockerhub/tag.js (hunk 7 above)
Comment (outloudvi) [reply lib/v2/dockerhub/tag.js:27]: Give nthat `link` is just the first architecture variant of the image:
* If the image didn't change but the order of the architecture changed, we might get different SHA256. Thus, the same tag push might give different hashes.
* If the image on a certain architecture changed but not the one we pick SHA256 on, Thus, a change might not be detected.

Probably we can collect all SHA256s and integrate them to a single GUID for the sake of robustness.

**Comment by TonyRL** [comment]
> > Consider adding an optional page_size parameter to the API endpoint as some image(OpenJDK) has a lot of tags.
> 
> Sounds great. By the way, is there any policy here on the paginator?

The upper bound of `page_size` 100 seems more than enough. But you can do it if you want. Just keep in mind that some images have quite a lot of pages that you may hit the API limit at some point.
Also, pulling data from 1 year ago is not what RSS means.

File: lib/v2/dockerhub/tag.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+const { parseDate } = require('@/utils/parse-date');
+const { hash } = require('./utils');
+
+module.exports = async (ctx) => {
+    const { owner, image, limits } = ctx.params;
+
+    const namespace = `${owner}/${image}`;
+    const link = `https://hub.docker.com/r/${namespace}`;
+
+    const pageSize = limits || 10;
```
Comment (TonyRL) [inline lib/v2/dockerhub/tag.js:11]:
```suggestion
    const pageSize = parseInt(limits) ?? 10;
```

File: docs/en/program-update.md [hunk 9]
```diff
@@ -129,8 +143,8 @@ The owner of the official image fills in the library, for example: https://rsshu
 Language
 
 | Simplified Chinese | Traditional Chinese | English |
-| ------ | -------- | -------- |
-|  cn  |  tw  |  en  |
+| ------------------ | ------------------- | ------- |
+| cn                 | tw                  | en      |
```
Comment (TonyRL) [inline docs/en/program-update.md:147]: Unnecessary changes?

File: docs/en/program-update.md [hunk 10]
```diff
@@ -296,8 +310,8 @@ Refer to [#minecraft](/en/game.html#minecraft)
 <RouteEn author="nczitzk" example="/rescuetime/release-notes" path="/rescuetime/release-notes/:os?" :paramsDesc="['OS id, see below']">
 
 | Mac OS | Windows |
-| - | - |
-| mac | windows |
+| ------ | ------- |
+| mac    | windows |
```
Comment (TonyRL) [inline docs/en/program-update.md:314]: Unnecessary changes?

**Review by TonyRL — CHANGES_REQUESTED** [review]
